filecoin-proofs/src/api/sector_builder/pieces.rs

```diff
@@ -0,0 +1,208 @@
+use sector_base::api::bytes_amount::{UnpaddedByteIndex, UnpaddedBytesAmount};
+use std::cmp::max;
+use std::io::Cursor;
+use std::io::Read;
+use std::iter::Iterator;
+
+pub struct PieceAlignment {
+    pub left_bytes: UnpaddedBytesAmount,
+    pub right_bytes: UnpaddedBytesAmount,
+}
+
+/**
+ * Given a list of pieces, sum the number of bytes taken by those pieces in that order.
+ */
+pub fn sum_piece_bytes_with_alignment(pieces: &[PieceMetadata]) -> UnpaddedBytesAmount {
+    pieces.iter().fold(UnpaddedBytesAmount(0), |acc, p| {
+        let PieceAlignment {
+            left_bytes,
+            right_bytes,
+        } = get_piece_alignment(acc, p.num_bytes);
+
+        acc + left_bytes + p.num_bytes + right_bytes
+    })
+}
+
+/**
+ * Given a list of pieces, find the piece with a matching piece_key.
+ */
+pub fn get_piece_by_key(pieces: &[PieceMetadata], piece_key: &str) -> Option<PieceMetadata> {
+    pieces
+        .iter()
+        .find(|p| p.piece_key == piece_key)
+        .map(PieceMetadata::clone)
+}
+
+/**
+ * Given a list of pieces, find the byte where a given piece does or would start.
+ */
+pub fn get_piece_start_byte(pieces: &[PieceMetadata], piece: &PieceMetadata) -> UnpaddedByteIndex {
+    // get pieces up to the target piece, or all of them if the piece doesn't exist
+    let pieces: Vec<PieceMetadata> = pieces
+        .iter()
+        .take_while(|p| p.piece_key != piece.piece_key)
+        .map(PieceMetadata::clone)
+        .collect();
+    // sum up all the bytes taken by the ordered pieces
+    let last_byte = sum_piece_bytes_with_alignment(&pieces);
+    let alignment = get_piece_alignment(last_byte, piece.num_bytes);
+
+    // add only the left padding of the target piece to give the start of that piece's data
+    UnpaddedByteIndex::from(last_byte + alignment.left_bytes)
+}
+
+/**
+ * Given a number of bytes already written to a staged sector (ignoring bit padding) and a number
+ * of bytes (before bit padding) to be added, return the alignment required to create a piece where
+ * len(piece) == len(sector size)/(2^n) and sufficient left padding to ensure simple merkle proof
+ * construction.
+ */
+pub fn get_piece_alignment(
+    written_bytes: UnpaddedBytesAmount,
+    piece_bytes: UnpaddedBytesAmount,
+) -> PieceAlignment {
+    // Bit padding causes bytes to only be aligned at every 127 bytes (for 31.75 bytes).
+    // @TODO change this away from magic numbers.
+    let minimum_piece_bytes = (4 * 32) - 1;
+    // Ensure the piece will be at the calculated minimum.
+    let adjusted_piece_bytes = max(minimum_piece_bytes, u64::from(piece_bytes));
+
+    let mut piece_bytes_needed = minimum_piece_bytes;
+
+    // Calculate the next power of two multiple that will fully contain the piece's data.
+    // This is required to ensure a clean piece merkle root, without being affected by
+    // preceding or following pieces.
+    while piece_bytes_needed < adjusted_piece_bytes {
+        piece_bytes_needed *= 2;
+    }
```

Comment:
BLOCKING

Suppose we replace `adjusted_piece_bytes` with just `piece_bytes` (converted to `u64`).

In the case that `piece_bytes` is greater than `minimum_piece_bytes`, then `adjusted_piece_bytes` is equal to the value of `piece_bytes` and the suggested replacement will not change the meaning.

If instead, `piece_bytes` is less than `minimum_piece_bytes`, then the code in the loop will never run. In that case, the final value of `piece_bytes_needed` will be `minimum_piece_bytes` — the same as the current implementation yields.

For that reason, I think we can completely remove the assignment of `adjusted_piece_bytes` and just use `piece_bytes`. If I haven't misunderstood, then doing this would simplify what needs to be understood to follow the algorithm being applied.